IFC4x3/Sections/Core data schemas/Schemas/IfcProductExtension/Entities/IfcIntegerVoxelData/DocEntity.xml

```diff
@@ -1,12 +1,34 @@
 <?xml version="1.0" encoding="utf-8"?>
 <DocEntity xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="IfcIntegerVoxelData" UniqueId="f3d953cb-24b7-483d-85e1-d44c43d4d22b" BaseDefinition="IfcVoxelData" EntityFlags="32">
 	<Attributes>
-		<DocAttribute Name="Values" UniqueId="342aa966-eaa4-4d0c-87e5-91988a4f1c0a" DefinedType="IfcInteger" AggregationType="2" AggregationLower="1">
+		<DocAttribute Name="Values" UniqueId="342aa966-eaa4-4d0c-87e5-91988a4f1c0a" DefinedType="IfcInteger" AggregationType="2" AggregationLower="1" AggregationUpper="GridSize">
 			<Documentation>The values assigned to the voxels. First x, then y and lastly z.</Documentation>
 		</DocAttribute>
 		<DocAttribute Name="Unit" UniqueId="d23d69e2-c35a-41d5-bf3b-9533670b9eb7" DefinedType="IfcUnit" AttributeFlags="1">
 			<Documentation>An optional unit for the integer values.</Documentation>
 		</DocAttribute>
+		<DocAttribute Name="GridSize" UniqueId="6f8d6a89-e211-4ca1-a678-e4221b561f42" DefinedType="IfcInteger">
+			<Derived> SIZEOF(SELF\IfcProduct.Representation.Representations[1].Items[1]\IfcVoxelGrid.Voxels)</Derived>
```

Comment:
1. There are a lot of `[1]` getters. The `GridSize` derive would fail with:

- two `Representations`, and
- `IfcVoxelGrid` being in the second, 

although the information is provided. 

2. `IfcRepresentation.Items` is a SET - accessing element at `[1]` doesn't work.

I believe, the expression should be refactored.

Reply:
> You were probably aiming at fetching from a SET with two member though, right? In that case I understand.

Correct.

> The only other option I can think of is a REPEAT to iterate over representations and find the one with IfcVoxelGrid.

This seems a reasonable solution.

> Your comments are not taking into account the VoxelGridRepresentation WR. I could also restrict SIZEOF Items to 1 and the [1] item to IfcVoxelGrid.

The current state of the expression is:

    'IFCVOXELGRID' IN TYPEOF(SELF\IfcProduct.Representation.Representations[1].Items[1])

Expanding it to ensure size to be 1 would solve the issue; however, I believe that limiting to a single representation would be very restrictive. Is the requirement "there should be one representation of type `IfcVoxelGrid`" or "there should be *only* one representation of type `IfcVoxelGrid`"?

Reply:
> Expanding it to ensure size to be 1 would solve the issue; however, I believe that limiting to a single representation would be very restrictive. Is the requirement "there should be one representation of type `IfcVoxelGrid`" or "there should be _only_ one representation of type `IfcVoxelGrid`"?

To me it would feel logical that an IfcVoxelData instance should have exactly one representation and it should be of type IfcVoxelGrid. On the other hand, there could be several IfcVoxelData sharing the same IfcVoxelGrid.

Reply:
> an IfcVoxelData instance should have exactly one representation and it should be of type IfcVoxelGrid

I can live with that. We can include this in #618 as well (e.g. see [here](https://github.com/bSI-InfraRoom/IFC-Specification/blob/f8e8310e0ca8bc19b50ac40f4bf8b37146494c2b/IFC4x3/Sections/Resource%20definition%20data%20schemas/Schemas/IfcRepresentationResource/Functions/IfcShapeRepresentationTypes/Expression.txt#L191-L198)).

Reply:
To me this derived attribute is more evidence that voxels as used in 4.4 shouldn't be burried deeply into the representation item, but should really be in the property domain. If these kind of direct dependencies are added it's clear that this is not a representation in line of how representations in IFC are currently used.

`set`s can be indexed numerically, but the only guarantee is that the indexing is stable. Index `n` remains index `n` for the lifetime of the `set`.

--

I also don't really understand the meaning of this attribute. In the IfcVoxelGrid there is `VoxelSizeX`,  `NumberOfVoxelsX` and `Voxels`. The last one is a 1-d flat sequence of booleans does that really make sense to expose upwards? It seems to me like the least meaningful attribute for an end-user?

Reply:
> To me this derived attribute is more evidence that voxels as used in 4.4 shouldn't be burried deeply into the representation item, but should really be in the property domain.

Voxels are used as geometry items in software. We also have presentation requirements for them, so the definition that fits most is _IfcGeometricRepresentationItem_.

> I also don't really understand the meaning of this attribute. In the IfcVoxelGrid there is `VoxelSizeX`, `NumberOfVoxelsX` and `Voxels`.

It is in the same ball park as Dim. Meta information that is derived from somewhere else. I do agree that extracting it from geometry is unorthodox but I would like to see geometry used more extensively for generating information from the model.

Reply:
One more thing. I'm not an expert at IfcDoc definitions, but does this actually result in an OPTIONAL  element specification?

```
Values : ARRAY [1:GridSize] OF OPTIONAL IfcInteger;
                               ^^^^^^^^
```

Reply:
And yet one more thing. This only fixes IfcIntegerVoxelData, but there are more right?

Reply:
Yes, before merging, we would need to address all other subtypes of `IfcVoxelData` also.